blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/MetricIndexData.java

```diff
@@ -0,0 +1,204 @@
+ * we get this below response from ES where key is the metric index and doc_count is
+ * the number of documents(think metric names) it points to.
+ *
+ * "buckets" : [ {
+ *     "key" : "foo.bar",
+ *     "doc_count" : 2
+ *   }, {
+ *     "key" : "foo.bar.baz",
+ *     "doc_count" : 2
+ *   }, {
+ *     "key" : "foo.bar.baz.aux",
+ *     "doc_count" : 1
+ * } ]
+ *
+ * If we feed each metric index to this class by calling the add(metricIndex, docCount) method,
+ * this class analyzes the data and provides methods to retrieve different aspects of this data
+ * with respect to baseLevel(if set to 2 with above data, uses foo.bar as reference).
+ *
+ */
+public class MetricIndexData {
+
+    private static class MetricIndexDocCount {
+
+        private long actualDocCount;
+        private long childrenTotalDocCount;
+
+        public MetricIndexDocCount(long actualDocCount, long childrenTotalDocCount) {
+            this.actualDocCount = actualDocCount;
+            this.childrenTotalDocCount = childrenTotalDocCount;
+        }
+
+        public void setActualDocCount(long actualDocCount) {
+            this.actualDocCount = actualDocCount;
+        }
+
+        public void addChildrenTotalDocCount(long childrenDocCount) {
+            this.childrenTotalDocCount += childrenDocCount;
+        }
+    }
+
+    private final int baseLevel;
+
+    //contains next level of tokens after baseLevel, which also have subsequent next level
+    private final Set<String> tokensWithNextLevelSet = new LinkedHashSet<String>();
+
+    //contains all metric indexes which are of the same length as baseLevel
+    final Map<String, MetricIndexDocCount> metricIndexSameLevelMap = new HashMap<String, MetricIndexDocCount>();
+
+    //contains all metric indexes which are more than the length of baseLevel by one.
+    final Map<String, MetricIndexDocCount> metricIndexNextLevelMap = new HashMap<String, MetricIndexDocCount>();
+
+    public MetricIndexData(int baseLevel) {
+        this.baseLevel = baseLevel;
+    }
+
+    /**
+     * For a given metricIndex and docCount, classifies the data with respect to baseLevel
+     * and stores it accordingly.
+     *
+     * @param metricIndex
+     * @param docCount is document count
+     */
+    public void add(String metricIndex, long docCount) {
+
+        final String[] tokens = metricIndex.split(AbstractElasticIO.REGEX_TOKEN_DELIMTER);
+
+        /**
+         *
+         * For the ES response shown in class description, for a baseLevel of 2,
+         * the data is classified as follows
+         *
+         * tokensWithNextLevelSet    -> {bar}               (tokens after baseLevel which also have a subsequent level)
+         *
+         * Both these maps maintain data in the form {metricIndex, (actualDocCount, childrenTotalDocCount)}
+         *
+         * metricIndexSameLevelMap   -> {foo.bar -> (2, 2)}     (all indexes which are of same length as baseLevel)
```

Comment:
OK, i guess with the docCount being a sum of all paths, including grandchildren, that's where I've gotten confused.